### PR TITLE
Extend pipeline to combine results in one tsv file

### LINKED ISSUE #15: Combine MotEvo results into one TSV file
**Is your feature request related to a problem? Please describe.**
Currently we have a separate results directory for every analysed PWM (as the output of parallell MotEvo runs). It would be nice to gather these results together into a nice table in TSV format.

**Describe the solution you'd like**
Extend the pipeline with one more step at the end: `combine_MotEvo_results`

### PR DESCRIPTION
**Pull request by krish8484**
## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

All the motevo results are encapsulated in one tsv file

Fixes #15 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not reduced the existing code coverage

### CODE REVIEW

File: tests/integration/expected_output.md5 [hunk 1]
```diff
@@ -20,3 +20,5 @@ fe56b8577bada06fea45bb8835e7d7ee ./tests/integration/output/motevo/motif_HNRNPF_
 faa2f9e3e8f432575e281878d93a85a6 ./tests/integration/output/motevo/motif_HNRNPF_824/posteriors
 c0a4a9155ec7958b00bd3140e97c6dc9 ./tests/integration/output/motevo/motif_HNRNPF_823/posterior_sites
 20e7921795665bbe24e585e741705634 ./tests/integration/output/motevo/motif_HNRNPF_823/posteriors
+bb5480584d2aba7c8cfae0d5629deba9 ./tests/integration/output/combined_MotEvo_results.tsv
+0f62e3b0dc5bc8b048a021632ace92fe ./tests/integration/output/combined_MotEvo_results.csv
```
Comment (AngryMaciek) [inline tests/integration/expected_output.md5:24]: There is also a corresponding `expected_output.txt` with the list of files, you should update that one as well.

File: workflow/Snakefile [hunk 2]
```diff
@@ -37,16 +37,13 @@ rule all:
...
     input:
-        PWM_input_dir = config["pwm_directory"],
-        DIR_MotEvo_results_pwm = expand(
-            os.path.join(
-                "{outdir}",
-                "motevo",
-                "{pwm_dir}"
-            ),
-            outdir = config["outdir"],
-            pwm_dir = gather_motifs_names(config["pwm_directory"])
```
Comment (AngryMaciek) [inline workflow/Snakefile:48]: As this is the input to your rule we do not need it in the last rule of the workflow anymore.

File: workflow/Snakefile [hunk 3]
```diff
@@ -282,3 +279,24 @@ rule run_MotEvo_analysis:
...
+        )
+    output:
+        combined_MotEvo_results = expand(os.path.join(
+            "{outdir}",
+            "combined_MotEvo_results.tsv"
+        ),
+        outdir = config["outdir"]
+        ),
+    params: "posterior_sites"
+    script: "parse.py"
```
Comment (AngryMaciek) [inline workflow/Snakefile:302]: Now I realised that if we call it like this then the script depends on the content of the `snakemake.`. This is bad, as we would not be able to call it outside of the pipeline ever (implying no unit tests as well).
Please change it to the `bash` call as for the other rules.

File: workflow/parse.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+ 
+tabb = dict()
+tabb['strand'] = []
+tabb['binding_position'] = []
+tabb['binding_posterior'] = []
+tabb['pwm_id'] = []
+tabb['binding_region'] = []
+tabb['binding_sequence'] = []
+tabb['binding_energy'] = []
+tabb['fast_record'] =[]
```
Comment (AngryMaciek) [inline workflow/parse.py:13]: the format is called `FASTA`, it would be better to add the `a` and we have the full name here.

File: workflow/parse.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
+import os
```
Comment (AngryMaciek) [inline workflow/parse.py:1]: Please add your authorship and License (Apache 2.0) on top, you may look it up in one of my other scripts to notice how I do it.

File: workflow/parse.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+tabb['binding_position'] = []
+tabb['binding_posterior'] = []
+tabb['pwm_id'] = []
+tabb['binding_region'] = []
+tabb['binding_sequence'] = []
+tabb['binding_energy'] = []
+tabb['fast_record'] =[] 
+
+print(snakemake.input)
+print(snakemake.params)
```
Comment (AngryMaciek) [inline workflow/parse.py:16]: There will not be here, as in the previous comment - we have to make a script independent of the workflow.

File: workflow/parse.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+filenameparam = str(snakemake.params)
+
+dirrs = snakemake.input
+outdir = str(snakemake.output)
+print(outdir)
```
Comment (AngryMaciek) [inline workflow/parse.py:22]: Please remove all such `print` checks in the final version. We want the code clean.

File: workflow/parse.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+for dirr in dirrs:  
+    abs_path = os.path.join(dirr, filenameparam) 
+    print(abs_path)
+
+    filename = os.path.relpath(abs_path)
+    print(filename)
+
+    i = 0
+    j = 0
+    file = open(filename, "r")
```
Comment (AngryMaciek) [inline workflow/parse.py:38]: Your code is not commented at all. Granted, these are trivial things but here you could inject 2,3 lines of comment to say how do you iterate over what et cetera.

File: workflow/Snakefile (hunk 3 above)
Comment (AngryMaciek) [reply workflow/Snakefile:302]: You will have to pass arguments to the script explicitly, please take a look at the file `format-ATtRACT-motifs.py` to see how to catch CLI arguments from Python.

File: workflow/Snakefile [hunk 9]
```diff
@@ -282,3 +279,24 @@ rule run_MotEvo_analysis:
         {input.DIR_pwm}) \
         1> {log.LOG_local_stdout} 2> {log.LOG_local_stderr}
         """
+
+rule combine_MotEvo_results:
+    input:
```
Comment (AngryMaciek) [inline workflow/Snakefile:284]: You are missing a rule description, check out the other rules in the workflow.

**Comment by AngryMaciek** [comment]
@krish8484 
Apart from what is in the code comments, I have unchecked some boxes from the checklist that still need to be addressed. As you change how the script is called you will also have to add unit tests for your script (the same way I have for another one). In order to generate dummy input for the unit test just run the pipeline once, copy the main directory with MotEvo results (per motif directories) and remove all of them except, let's say, arbitrarily chosen 10 (so that the test is rather small).

**Comment by krish8484** [comment]
Okay, i had done it the wrong way, will fix it thanks!

**Comment by krish8484** [comment]
All the changed mentioned above have been implemented. Both the `md5sum --check` commands are returning `OK` on my local machine, but failing on the travis server. I am not understanding the reason.

File: .travis.yml [hunk 10]
```diff
@@ -48,3 +48,10 @@ script:
   # Test pipeline execution: local, conda envs:
   - bash tests/integration/execution/snakemake_local_run_conda_environments.sh
   - md5sum --check tests/integration/expected_output.md5
+  # Test parse.py script
+  - >
+    python scripts/parse.py
+    --indir tests/unit/combine_results/motif_HNRNPF_820 tests/unit/combine_results/motif_HNRNPF_821 tests/unit/combine_results/motif_HNRNPF_822 tests/unit/combine_results/motif_HNRNPF_823 tests/unit/combine_results/motif_HNRNPF_824
+    --filename posterior_sites
+    --outfile tests/unit/combine_results/combined_MotEvo_results.tsv
+  - md5sum --check tests/unit/combine_results/expected_output.md5
```
Comment (AngryMaciek) [inline .travis.yml:57]: This is OK, but please put it before the integration test. As a good practice: All unit tests should be executed and passed before the integration test starts.

File: envs/main.yml [hunk 11]
```diff
@@ -28,5 +28,4 @@
     - pyyaml=3.12
     - snakemake=5.19.0
     - unzip=6.0
-
```
Comment (AngryMaciek) [inline envs/main.yml:31]: If the other YAML files have space here then please leave it as it was, for consistency.

File: scripts/parse.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
+##############################################################################
+#   AUTHOR: Krish Agarwal
+#   AFFILIATION: University_of_Basel
+#   AFFILIATION: Swiss_Institute_of_Bioinformatics
```
Comment (AngryMaciek) [inline scripts/parse.py:4]: As you are not officially an SIB member I would remove this line but keep only University of Basel, as you are working in our research group now.

File: tests/integration/output/local_log/prepare_MotEvo_parameters.benchmark.log [hunk 13]
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
Comment (AngryMaciek) [inline tests/integration/output/local_log/prepare_MotEvo_parameters.benchmark.log:1]: This is a log file with execution time - it should not be under version control. Miliseconds of execution might change between pipeline runs but that is OK.

Comment (AngryMaciek) [reply tests/integration/output/local_log/prepare_MotEvo_parameters.benchmark.log:1]: The same applies to two streams below: empty `stderr` and `stdout`

File: tests/integration/output/local_log/prepare_sequence_for_MotEvo.benchmark.log [hunk 14]
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
Comment (AngryMaciek) [inline tests/integration/output/local_log/prepare_sequence_for_MotEvo.benchmark.log:1]: The same situation here - this is a log file

File: tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_820.benchmark.log [hunk 15]
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
Comment (AngryMaciek) [inline tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_820.benchmark.log:1]: The same situation here - this is a log file

File: tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_821.benchmark.log [hunk 16]
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
Comment (AngryMaciek) [inline tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_821.benchmark.log:1]: The same situation here - this is a log file

File: tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_822.benchmark.log [hunk 17]
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
Comment (AngryMaciek) [inline tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_822.benchmark.log:1]: The same situation here - this is a log file.
This applies to all the `stderr` and `stdout` files as well(!)

File: tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_823.benchmark.log [hunk 18]
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
Comment (AngryMaciek) [inline tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_823.benchmark.log:1]: The same situation here - this is a log file

File: tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_824.benchmark.log [hunk 19]
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
Comment (AngryMaciek) [inline tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_824.benchmark.log:1]: The same situation here - this is a log file

File: tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_825.benchmark.log [hunk 20]
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
Comment (AngryMaciek) [inline tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_825.benchmark.log:1]: The same situation here - this is a log file

File: tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_826.benchmark.log [hunk 21]
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
+0.0984	0:00:00	3.59	10.85	0.18	0.78	0.00	0.00	0.00
```
Comment (AngryMaciek) [inline tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_826.benchmark.log:2]: The same situation here - this is a log file

File: tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_827.benchmark.log [hunk 22]
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
Comment (AngryMaciek) [inline tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_827.benchmark.log:1]: The same situation here - this is a log file

File: tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_828.benchmark.log [hunk 23]
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
Comment (AngryMaciek) [inline tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_828.benchmark.log:1]: The same situation here - this is a log file

File: tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_829.benchmark.log [hunk 24]
```diff
@@ -0,0 +1,2 @@
+s	h:m:s	max_rss	max_vms	max_uss	max_pss	io_in	io_out	mean_load
```
Comment (AngryMaciek) [inline tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_829.benchmark.log:1]: The same situation here - this is a log file

File: workflow/Snakefile (hunk 9 above)
Comment (AngryMaciek) [reply workflow/Snakefile:287]: This is OK now,

**Comment by AngryMaciek** [comment]
I have made some minor comments, please remove unnecessary files first, before I git into the script itself.
Also, please remove all `.snakemake_timestamp ` files. There are internal time stamps that should not be under version control;

File: tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_820.stdout.log [hunk 25]
```diff
@@ -0,0 +1,53 @@
+letter A val is 0.250000
```
Comment (AngryMaciek) [inline tests/integration/output/local_log/run_MotEvo_analysis.motif_HNRNPF_820.stdout.log:1]: Please remove ALL `stdout` and `stderr` files from the repository. These are logs and should not be kept under version control.  
It makes me wonder - why did you add these exact ones in the first place? These streams are related to `run_MotEvo_analysis` rule, which I already covered...

File: tests/integration/output/sequence.fasta [hunk 26]
```diff
@@ -0,0 +1,2 @@
+>>MOTEVO_input-sequence
+ATGTGAGTGAAGTGTGGGAAAGATGACTCGATATATCTGGATGCTAGGGATCGGATGGCGATACG
```
Comment (AngryMaciek) [inline tests/integration/output/sequence.fasta:2]: Why do you need this file under version control of the integration test?

File: workflow/Snakefile [hunk 27]
```diff
@@ -282,3 +279,29 @@ rule run_MotEvo_analysis:
...
+        combined_MotEvo_results = expand(os.path.join(
+            "{outdir}",
+            "combined_MotEvo_results.tsv"
+        ),
+        outdir = config["outdir"]
+        ),
+    params:
+        Input_filename = "posterior_sites"
+    shell: 
+        "python ../../../scripts/parse.py --indir {input.DIR_MotEvo_results_pwm}  --filename {params.Input_filename} --outfile {output.combined_MotEvo_results}"
```
Comment (AngryMaciek) [inline workflow/Snakefile:307]: Please break the line with `\` characterl (as in the other rule); there is a line width limit

**Comment by krish8484** [comment]
The log files were auto generated after running the pipeline, you are right they were not needed, so I deleted them now.
The tests are passing sometime and failing other time because the order of input directories that the snakemake file receives is not the same all the time, so the order of tsv file changes according to the input.
I have implemented the other changes.

**Comment by AngryMaciek** [comment]
> The integration tests are passing sometime and failing other time because the order of input directories that the snakemake file receives is not the same all the time, so the order of tsv file changes according to the input. I think unit tests do a good job of checking the script.

OK, thanks! Looks good now. But before I merge we need to take care about this randomness effect. Reproducibility is a very important aspect of data analysis and we cannot leave any non-deterministic effects like that. That is a easy fix, though - just sort the list of paths to the files with `sorted()` buitl-in function. Then the order will always be the same.

File: .travis.yml [hunk 28]
```diff
@@ -42,9 +42,17 @@ script:
     --organism Mus_musculus
     --outdir tests/unit/format-ATtRACT-motifs/ATtRACT_mmu
   - md5sum --check tests/unit/format-ATtRACT-motifs/expected_output_mmu.md5
+  # Test parse.py script
+  - >
+    python scripts/parse.py
+    --indir tests/unit/combine_results/motif_HNRNPF_820 tests/unit/combine_results/motif_HNRNPF_821 tests/unit/combine_results/motif_HNRNPF_822 tests/unit/combine_results/motif_HNRNPF_823 tests/unit/combine_results/motif_HNRNPF_824
```
Comment (AngryMaciek) [inline .travis.yml:48]: could you please change the argument to `input-directories` explicite? :)

File: scripts/parse.py [hunk 29]
```diff
@@ -0,0 +1,111 @@
...
+df = pd.DataFrame({key: pd.Series(value) for key, value in tabb.items()})
+df.to_csv(rel_outfile_csv, index=False)
+
+
+with open(rel_outfile_csv, 'r') as csvin, open(rel_outfile, 'w') as tsvout:
+    csvin = csv.reader(csvin)
+    tsvout = csv.writer(tsvout, delimiter='\t')
+
+    for row in csvin:
+        tsvout.writerow(row) #printing row by row in tsv
```
Comment (AngryMaciek) [inline scripts/parse.py:107]: We do not need two output files, just the `tsv` is enough.
You may use: `df.to_csv(rel_outfile_csv, index=False, sep="\t)`

Comment (AngryMaciek) [reply scripts/parse.py:107]: This applies to all the CI - you should check just the `tsv file`

File: scripts/parse.py [hunk 30]
```diff
@@ -0,0 +1,111 @@
...
+for dirr in dirrs:
+    abs_path = os.path.join(dirr, filenameparam)
+
+    filename = os.path.relpath(abs_path)
+
+    i = 0
+    j = 0
+    file = open(filename, "r")
+    for each in file:
+        each_word = each.split(' ')
```
Comment (AngryMaciek) [inline scripts/parse.py:85]: You still do not have any comments on your code...

File: tests/unit/combine_results/combined_MotEvo_results.csv [hunk 31]
```diff
@@ -0,0 +1,22 @@
+strand,binding_position,binding_posterior,pwm_id,binding_region,binding_sequence,binding_energy,fasta_record
++,14-18,0.015291,HNRNPF_820,"MOTEVO_input-sequence
+",GTGGG,2.51049,"MOTEVO_input-sequence
```
Comment (AngryMaciek) [inline tests/unit/combine_results/combined_MotEvo_results.csv:3]: I am very confused by the information here... why is every record separated into two rows? Each record should be one line (i.e. one row) in the result table. The records should match the header of the table.

File: tests/unit/combine_results/combined_MotEvo_results.tsv [hunk 32]
```diff
@@ -0,0 +1,22 @@
...
+"	AGGGA	6.71171	"MOTEVO_input-sequence
+"
++	15-19	0.560144	HNRNPF_824	"MOTEVO_input-sequence
+"	TGGGA	6.71171	"MOTEVO_input-sequence
+"
++	46-50	0.030141	HNRNPF_824	"MOTEVO_input-sequence
+"	AGGGA	2.51049	"MOTEVO_input-sequence
+"
++	51-55	0.014195	HNRNPF_824	"MOTEVO_input-sequence
+"	TCGGA	2.51049	"MOTEVO_input-sequence
```
Comment (AngryMaciek) [inline tests/unit/combine_results/combined_MotEvo_results.tsv:21]: I am very confused by the information here... why is every record separated into two rows? Each record should be one line (i.e. one row) in the result table. The records should match the header of the table.

File: tests/unit/combine_results/combined_MotEvo_results.tsv [hunk 33]
```diff
@@ -0,0 +1,22 @@
+strand	binding_position	binding_posterior	pwm_id	binding_region	binding_sequence	binding_energy	fasta_record
```
Comment (AngryMaciek) [inline tests/unit/combine_results/combined_MotEvo_results.tsv:1]: I would rearrange the order of columns here a little:
`pwm_id binding_position, binding_sequence	binding_posterior	binding_energy`

We do not need `strand` or `fasta_record` for now. Please keep them in the code but just do not use them to generate the table. The same applies to `binding_region` - note that it is the same as `fasta_record`

Comment (krish8484) [reply tests/unit/combine_results/combined_MotEvo_results.tsv]: Done!

File: tests/unit/combine_results/combined_MotEvo_results.tsv (hunk 32 above)
Comment (krish8484) [reply tests/unit/combine_results/combined_MotEvo_results.tsv]: That is because you are viewing the tsv file in GitHub. Have you checked it using LibreOffice?

File: scripts/parse.py (hunk 30 above)
Comment (krish8484) [reply scripts/parse.py]: Added!

File: scripts/parse.py (hunk 29 above)
Comment (krish8484) [reply scripts/parse.py]: Done!

File: .travis.yml (hunk 28 above)
Comment (krish8484) [reply .travis.yml]: Done!

**Comment by AngryMaciek** [comment]
OK, looks very good now. ╰(▔∀▔)╯  
  
As I was going through the files in detail I also noticed that you do not have a *conda* virtual environment for your rule, right? Each snakemake rule is executed in a separated, dedicated virtual environment which contains all the packages needed. They are automacitally built by snakemake during the pipeline execution, built based on the YAML files. Take a look at other rules in the workflow. Could you please add a YAML file with specifications related to your script and a proper field in the Snakefile?

File: tests/unit/combine_results/combined_MotEvo_results.tsv (hunk 32 above)
Comment (AngryMaciek) [reply tests/unit/combine_results/combined_MotEvo_results.tsv]: Oh, OK, thanks! GitHub confused me...

File: workflow/Snakefile [hunk 34]
```diff
@@ -282,3 +279,32 @@ rule run_MotEvo_analysis:
...
+        combined_MotEvo_results = expand(os.path.join(
+            "{outdir}",
+            "combined_MotEvo_results.tsv"
+        ),
+        outdir = config["outdir"]
+        ),
+    params:
+        Input_filename = "posterior_sites"
+    shell: 
+        "python ../../../scripts/parse.py \
```
Comment (AngryMaciek) [inline workflow/Snakefile:307]: This is a comment for @AngryMaciek :
Double check how to encode relative paths, also - check the execution bash scripts for `cwd` there

**Comment by AngryMaciek** [comment]
Please also run a `rulegraph run` test to generate new image of the workflow that will be displayed in the README (it should contain your rule). The new image file should substitute the old one in `images` directory.

File: .travis.yml [hunk 35]
```diff
@@ -42,9 +42,17 @@ script:
     --organism Mus_musculus
     --outdir tests/unit/format-ATtRACT-motifs/ATtRACT_mmu
   - md5sum --check tests/unit/format-ATtRACT-motifs/expected_output_mmu.md5
+  # Test parse.py script
+  - >
+    python scripts/parse.py
```
[the remark this answers is not in the document]
Comment (krish8484) [reply .travis.yml]: done, i have changed the path

**Comment by krish8484** [comment]
Updated the image of the workflow, added env file, formatted using `black` and changed the path of `parse.py`

File: workflow/Snakefile (hunk 34 above)
Comment (AngryMaciek) [reply workflow/Snakefile]: This is a more complex issue - I will have to create a separate feature branch for this fix.
